Framework/Core/README.md

```diff
@@ -265,40 +265,84 @@ the `ProcessingContext`  your computation  lambda is  passed and  contains one
 value for each of the `InputSpec` you specified. E.g.:
 
 ```cpp
-InputRecord &args = ctx.inputs();
+InputRecord &inputs = ctx.inputs();
 ```
 
 From the `InputRecord` instance you can get the arguments either via their positional
 index:
 
 ```cpp
-DataRef ref = args.getByPos(0);
+DataRef ref = inputs.getByPos(0);
 ```
 
 or using the mnemonics-label which was used as first argument in the associated
 `InputSpec`.
 
 ```cpp
-DataRef ref = args.get("points");
+DataRef ref = inputs.get("points");
 ```
 
 You can then use the `DataRef` `header` and `payload` raw pointers to access
 the data in the messages.
 
-If the message is of a known type, you can automatically get a a smart pointer
-to the contents of the message by passing it as template argument, e.g.:
+If the message is of a known messageable type, you can automatically get the content
+of the message by passing type T as template argument. The actual operation depends on
+the properties of the type. Not all types are supported, in order to get an object with
+pointer-like behavior, T has to be a pointer (T = U*).
+```cpp
+auto p = args.get<T>("input");
+```
+The return type is
+* ```T const&``` if T is a messageable type
+* ```T``` if T is a std::container of a ROOT-serializable type
+* ```smart_pointer<T>``` if T is a ROOT-serializable type and T* is passed
+
+Examples
+* for messageable types there is no additional copy involved, the content is only
+for reading.
+```cpp
+XYZ const& p = args.get<XYZ>("points");
+```
 
+* Root-serialized objects are automatically deserialized and returned as a smart
+pointer. Note that the requested type has to be pointer.
 ```cpp
-auto v = args.get<XYZ>("points");
-XYZ &p = *v;
+auto h = args.get<TH1*>("histo");
+h->Print();
 ```
 
-The framework will also take care of necessary deserialization.
+* container of Root-serialized objects are automatically deserialized and returned as
+container object.
+```cpp
+auto points = args.get<std::vector<TPoint>>("points");
+for (auto& point : points) {}
+```
+
+Check next  section  for known types.  The framework  will  also take  care of
+necessary deserialization.
 
 [InputRecord]: https://github.com/AliceO2Group/AliceO2/blob/HEAD/Framework/Core/include/Framework/InputRecord.h
 
-### Creating outputs - the DataAllocator API
+#### Utilities for working with `DataRef`
+Get payload size:
+```cpp
+size_t payloadSize = DataRefUtils::getPayloadSize(ref);
+```
 
+Extract a header from the header stack:
+```cpp
+const HeaderType* header = DataRefUtils::getHeader<HeaderType*>(ref);
```

Comment:
Just to understand it better: is it possible to put two headers of the same type onto stack and later get both of them?

Reply:
right now we have not put a limitation on that in the code, in principle one can add more than one header of the same type to the stack (I know that I have been proposing it to you). However, there are also arguments not to do so, and it's also written in the data model that there is only one `DataHeader` and the header message on the wire always starts with `DataHeader` It is probably confusing to have more than one, also extracting the second one requires some more coding. @mkrzewic, feel free to add your comment on that.

For the particular QC case one could think of a additional QCMetaHeader holding meta information in addition to the original DataHeader as a member. That can be send together with a QC data header in a stack.

Reply:
I also agree that there should be only one single header per type, just like we should have a single producer per data type at any given moment. BTW, at some point we discussed about putting some sort of timestamp / time range in the DataHeader. Can we resurrect that discussion? I thought we actually agreed that was a good idea.

Reply:
I am not opting too much for this solution as well. Having two of the same type would make it a bit unclear, which of them is more important or "the right one" in certain case. Also, some of the header fields would be duplicated (like the payload size). We can discuss adding some QC header. 

+1 for having the timestamp accessible somewhere in the user code, even if not necessarily in DataHeader. Actually, it will be a must for Data Sampling to access it somehow. A use case - sampling data of the same event, but different detectors (which will mean different FLPs and separate dispatchers).

Reply:
The timestamp is actually available from the `DataProcessingHeader` at the moment.

Reply:
@ktf Sorry, I meant `timesliceID`, not `timestamp` (I didn't realise it is a different thing). Is it accesible somewhere?

Reply:
```c++
    const DataProcessingHeader* dph = o2::header::get<DataProcessingHeader*>(header->GetData());
    assert(dph);
    size_t timesliceId = dph->startTime;
```

Reply:
or you mean the position in the ring buffer I use for keeping them in fly?

Reply:
Some unique ID of message which helps you to determine, that 2 given inputs of data processor correspond to each other (that they form a complete set of payloads). So this is `timesliceID` or sth else?

Reply:
yes, then you want the one above.

Reply:
did I answer your question?

Reply:
@ktf Yes, thanks!